### PR TITLE
Gitbook renderer

### PR DESCRIPTION
**Pull request by geramirez**
Gitbook renderer and tests
Test coverage is still hanging around 60% because I haven't tested methods that deal with I/O yet. However, the individual exports are ready.

### CODE REVIEW

File: models/certifications.go [hunk 1]
```diff
@@ -13,6 +14,20 @@ type Certification struct {
 	Standards map[string]Standard `yaml:"standards" json:"standards"`
 }
 
+// YieldAll returns a list of sorted standards
+func (certification Certification) GetSortedData(callback func(string, string)) {
+	var keys []string
```
Comment (jcscottiii) [inline models/certifications.go]: standardNames?

File: models/certifications.go [hunk 2]
```diff
@@ -13,6 +14,20 @@ type Certification struct {
...
+	for key := range certification.Standards {
```
Comment (jcscottiii) [inline models/certifications.go]: standardName

File: models/certifications.go [hunk 3]
```diff
@@ -13,6 +14,20 @@ type Certification struct {
...
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+	for _, standardKey := range keys {
+		certification.Standards[standardKey].GetSortedData(func(controlKey string) {
```
Comment (jcscottiii) [inline models/certifications.go]: controlName and standardName

File: models/components.go [hunk 4]
```diff
@@ -10,44 +10,25 @@ import (
...
+type SatisfiesList []Satisfies
+
 // Component struct is an individual component requiring documentation
 type Component struct {
 	Name          string                  `yaml:"name" json:"name"`
 	Key           string                  `yaml:"key" json:"key"`
-	References    []GeneralReference      `yaml:"references" json:"references"`
-	Verifications []VerificationReference `yaml:"verifications" json:"verifications"`
-	Satisfies     []Satisfies             `yaml:"satisfies" json:"satisfies"`
+	References    *GeneralReferences      `yaml:"references" json:"references"`
```
Comment (jcscottiii) [inline models/components.go]: just wondering. do these have to be pointers?

File: models/components.go [hunk 5]
```diff
@@ -76,6 +57,13 @@ func (components *Components) Get(key string) *Component {
 	return components.mapping[key]
 }
 
+//GetAndApply get a component and apply the
```
Comment (jcscottiii) [inline models/components.go]: Finish the comment

File: models/components.go [hunk 6]
```diff
@@ -76,6 +57,13 @@ func (components *Components) Get(key string) *Component {
...
+func (components *Components) GetAndApply(key string, callback func(component *Component)) {
```
Comment (jcscottiii) [inline models/components.go]: i feel like the repetition of GetAndApply can be solved with a base concrete struct and interfaces. not 100% sure though

Comment (jcscottiii) [inline models/components.go]: key -> componentName

File: models/justifications.go [hunk 7]
```diff
@@ -2,38 +2,89 @@ package models
 
 import "sync"
 
+// Verification struct holds data for a specific coponent and verification
```
Comment (jcscottiii) [inline models/justifications.go]: typo

File: models/standards.go [hunk 8]
```diff
@@ -26,6 +27,18 @@ type Standards struct {
 	sync.RWMutex
 }
 
+// YieldAll returns a list of sorted certifications
+func (standard Standard) GetSortedData(callback func(string)) {
```
Comment (jcscottiii) [inline models/standards.go]: do we actually use this one?

Comment (jcscottiii) [inline models/standards.go]: nvm

**Comment by jcscottiii** [comment]
Hmm. I'm sure about those functions that call a callback now. Like GetSortedData could just return sorted data. it makes reading a little harder. We can discuss this tomorrow.
I think you do all these callbacks to keep stuff locked while diving deeper and locking more stuff along the way.

You could just instead directly lock everything you need
justifications.Lock()
components.Lock()
//do stuff
components.Unlock()
justifications.Unlock()


Maybe i'm wrong though. let's discuss.

File: models/components.go (hunk 4 above)
Comment (geramirez) [inline models/components.go]: Nope, but I figured calling the sort method on them would keep them sorted in one location.

**Comment by geramirez** [comment]
I think we could lock the components and justifications explicitly in that way, but I feel the need to maintain all the locking mechanism next to the struct they are locking to keep from forgetting later. I'm not tied to either approach though.

File: models/certifications_test.go [hunk 9]
```diff
@@ -20,8 +31,64 @@ func TestLoadCertification(t *testing.T) {
...
+	for _, example := range certificationTestErrors {
+		openControl := &OpenControl{}
+		actualError := openControl.LoadCertification(example.certificationFile)
+		if example.expectedError != actualError {
+			t.Errorf("Expected %s, Actual: %s", example.expectedError, actualError)
+		}
+	}
+}
+
+var standardOrderTests = []standardOrderTest{
```
Comment (jcscottiii) [inline models/certifications_test.go]: Comment about what each case is doing/ testing

File: models/justifications_test.go [hunk 10]
```diff
@@ -3,29 +3,107 @@ package models
...
-	mappings          []singleMapping
-	expectedStandards int
+	mappings      []singleMapping
+	expectedCount int
 }
 
-var justificationsTests = []justificationsTest{
-	{[]singleMapping{{"s", "c", "1"}, {"s", "c", "2"}, {"s", "c", "3"}}, 1},
-	{[]singleMapping{{"s1", "c", "1"}, {"s2", "c", "2"}, {"s3", "c", "3"}}, 3},
+var justificationsAddTests = []justificationsTest{
```
Comment (jcscottiii) [inline models/justifications_test.go]: Comment about what each case is doing/ testing

File: models/justifications_test.go [hunk 11]
```diff
@@ -3,29 +3,107 @@ package models
...
 		}
-		if example.expectedStandards != len(just.mapping) {
-			t.Errorf("Expected %d, Actual: %d", example.expectedStandards, len(just.mapping))
+		if example.expectedCount != len(just.mapping) {
+			t.Errorf("Expected %d, Actual: %d", example.expectedCount, len(just.mapping))
+		}
+	}
+}
+
+var justificationsGetTests = []justificationsTest{
```
Comment (jcscottiii) [inline models/justifications_test.go]: Comment about what each case is doing/ testing

File: models/justifications_test.go [hunk 12]
```diff
@@ -3,29 +3,107 @@ package models
...
+		}
+	}
+}
+
+type verificationsLessTest struct {
+	verifications Verifications
+	leftIsLess    bool
+}
+
+var verificationsLessTests = []verificationsLessTest{
```
Comment (jcscottiii) [inline models/justifications_test.go]: Comment about what each case is doing/ testing

File: models/references_test.go [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+	for _, example := range generalReferenceLenTests {
+		actualLength := example.references.Len()
+		if example.expectedLength != actualLength {
+			t.Errorf("Expected %d, Actual: %d", example.expectedLength, actualLength)
+		}
+	}
+}
+
+var verificationReferencesLenTests = []verificationReferencesLenTest{
+	{VerificationReferences{}, 0},
```
Comment (jcscottiii) [inline models/references_test.go]: Comment about what each case is doing/ testing

File: models/references_test.go [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+	for _, example := range verificationReferencesLenTests {
+		actualLength := example.references.Len()
+		if example.expectedLength != actualLength {
+			t.Errorf("Expected %d, Actual: %d", example.expectedLength, actualLength)
+		}
+	}
+}
+
+var generalReferencesLessTests = []generalReferencesLessTest{
+	{GeneralReferences{GeneralReference{Name: "1"}, GeneralReference{Name: "2"}}, true},
```
Comment (jcscottiii) [inline models/references_test.go]: Comment about what each case is doing/ testing

File: models/references_test.go [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+	for _, example := range generalReferencesLessTests {
+		actualLeftIsLess := example.references.Less(0, 1)
+		if example.leftIsLess != actualLeftIsLess {
+			t.Errorf("Expected %s, Actual: %s", actualLeftIsLess, actualLeftIsLess)
+		}
+	}
+}
+
+var verificationReferencesLessTests = []verificationReferencesLessTest{
+	{VerificationReferences{VerificationReference{Key: "", GeneralReference: GeneralReference{Name: "1"}}, VerificationReference{Key: "", GeneralReference: GeneralReference{Name: "2"}}}, true},
```
Comment (jcscottiii) [inline models/references_test.go]: Comment about what each case is doing/ testing

File: models/standards_test.go [hunk 16]
```diff
@@ -8,21 +8,92 @@ type standardsTest struct {
...
+	}
+}
+
+type controlOrderTest struct {
+	standard      Standard
+	expectedOrder string
+}
+
+var controlOrderTests = []controlOrderTest{
+	{
```
Comment (jcscottiii) [inline models/standards_test.go]: Comment about what each case is doing/ testing

**Comment by jcscottiii** [comment]
@geramirez this looks good! on the metro, i thought of a way to address concurrency but want to discuss in person. for another PR. this is clean and well documented. merged.

**Pull request merged**